### ISSUE TITLE
Investigating type git special remotes

### ISSUE THREAD

**Issue by bpoldrack** [datalad/datalad#6204]
I am investigating our suggested usage of type git special remotes in tests, help pages (like `create-sibling-gin`) and the handbook.
This relates to https://github.com/datalad-handbook/book/issues/779, https://github.com/datalad/datalad/issues/6199, https://github.com/datalad/datalad/issues/5144, https://github.com/datalad/datalad/issues/6195 and also somewhat https://github.com/datalad/datalad/issues/6184.

I'll use this issue to keep track of what I discover.

1. To correctly `initremote` such a special remote the respective git remote needs to already be an accessible, valid annex with a uuid. Otherwise the special remote will end up with no uuid and is unusable.

2. That invalidates the description in the handbook (https://handbook.datalad.org/en/latest/basics/101-139-gin.html#using-gin-as-a-data-source-behind-the-scenes), since it suggests to add a freshly created repo on GIN with `siblings add ... -as-common-datasource`, but such a new repo is a plain git repo. In addition that description doesn't account for the fact, the subsequent push needs to deal with unrelated histories.

3. The test `distribution.tests.test_siblings.test_as_common_datasource` is partially broken, since it creates two such special remotes, one of which is invalid, not autoenabled and not used anywhere in the test. The clone on which the `get` is tested has this:
```
❱ git cat-file blob "git-annex:remote.log"
 autoenable=true location=http://127.0.0.1:34537/ name=fresh-sr type=git timestamp=1637150426.567013487s
0116253d-d1a2-458b-b6ba-e69180149a2b autoenable=true location=http://127.0.0.1:34537/.git name=mike2 type=git timestamp=1637150360.267307216s
```
Hence, `fresh-sr` is invalid (no uuid) and `mike2` works correctly and is in fact autoenabled. Note, that the two end up using different URLs (once with and once w/o `.git`)

4. `git-annex-init` doesn't complain about such autoenabled special remote w/o uuid. It simply doesn't do anything. A manual `enableremote` afterwards reveals that annex thinks there's no special remote.

**Comment by bpoldrack** [comment]
One more confusing thing about that test (`distribution.tests.test_siblings.test_as_common_datasource`).
There's this line cloning a dataset: https://github.com/datalad/datalad/blob/02f6914ba856333f00ab26af97b6e51fb8792104/datalad/distribution/tests/test_siblings.py#L542

Dropping into debugger right there, gives me:
```
(Pdb) p url
'http://127.0.0.1:44847/'
```
and in that resulting clone, I see 
```
❱ cat .git/config

...

[remote "dl-test-remote"]
	url = http://127.0.0.1:44847/.git
	fetch = +refs/heads/*:refs/remotes/dl-test-remote/*
	annex-bare = false
	annex-uuid = 1ca7b51d-7788-4dc0-ac3f-55f6bfe054c6
```
However, if I proceed to manually do the same cloning `datalad clone http://127.0.0.1:44847/ test` (same venv, datalad being on `master`), I get:

```
❱ cat .git/config 

...

[remote "origin"]
	url = http://127.0.0.1:44847/.git
	fetch = +refs/heads/*:refs/remotes/origin/*
	annex-ignore = true
```

I wonder why that outcome is different. Could there be a side-effect to us not using the name `origin` in the tests (or to how we do that)?

**Comment by bpoldrack** [comment]
Ok, ignoring the above oddity for the moment, this is what seems to work WRT GIN, @mslw :

Let's make a dataset:
```
❱ datalad create test
❱ cd test
❱ echo bla > some
❱ datalad save
```
Now, make a sibling on GIN:

```
❱ DATALAD_CREDENTIAL_GIN_TOKEN=<your_token> datalad create-sibling-gin public-test-repo2 -s public-gin --access-protocol https-ssh --credential gin
create_sibling_gin(ok): [sibling repository 'public-gin' created at https://gin.g-node.org/bpoldrack/public-test-repo2]
configure-sibling(ok): . (sibling)
action summary:
  configure-sibling (ok: 1)
  create_sibling_gin (ok: 1)
```

We have a a repo on GIN now, but this is a plain git repository as of now with no related history yet. Hence, it's ignored by annex.
Remove that config:

```
❱ git config --unset-all remote.public-gin.annex-ignore
```

And use `push` once to achieve two things:
- make it an actual annex (the git-annex branch is pushed)
- make git-annex aware that it is an annex and has a uuid

Furthermore, this also pushes the actual state of our dataset incl. content, but this isn't the point really at this stage.

```
❱ datalad push --to public-gin
copy(ok): some (file) [to public-gin...]                                                                 
publish(ok): . (dataset) [refs/heads/git-annex->public-gin:refs/heads/git-annex a56b602..e23876f]        
publish(ok): . (dataset) [refs/heads/master->public-gin:refs/heads/master [new branch]]                  
action summary:                                                                                          
  copy (ok: 1)
  publish (ok: 2)
```

Now, we do have a proper annex enabled git remote:

```
❱ cat .git/config

... <snip> ...

[remote "public-gin"]
	url = https://gin.g-node.org/bpoldrack/public-test-repo2
	fetch = +refs/heads/*:refs/remotes/public-gin/*
	annexurl = git@gin.g-node.org:/bpoldrack/public-test-repo2.git
	pushurl = git@gin.g-node.org:/bpoldrack/public-test-repo2.git
	annex-uuid = 8abc0817-a367-42a8-ba19-db08a8ece63f
```

Once the remote looks like this, we can add the special remote:

```
❱ datalad siblings configure -s public-gin --as-common-datasrc gin-sr     
.: public-gin(+) [https://gin.g-node.org/bpoldrack/public-test-repo2 (git)]
```

And make it known to the GIN sibling in order to propagate it - push again:
```
❱ datalad push --to public-gin
publish(ok): . (dataset) [refs/heads/git-annex->public-gin:refs/heads/git-annex e23876f..b578983]        
action summary:                                                                                          
  publish (notneeded: 1, ok: 1)
```

------------
Let's test that.

Make another sibling on github and push:

```
❱ datalad create-sibling-github public-testrepo -s public-github --publish-depends public-gin          
create_sibling_github(ok): [sibling repository 'public-github' created at https://github.com/bpoldrack/public-testrepo]
[INFO   ] Configure additional publication dependency on "public-gin" 
configure-sibling(ok): . (sibling)
action summary:
  configure-sibling (ok: 1)
  create_sibling_github (ok: 1)
❱ datalad push --to public-github
publish(ok): . (dataset) [refs/heads/master->public-github:refs/heads/master [new branch]]               
publish(ok): . (dataset) [refs/heads/git-annex->public-github:refs/heads/git-annex [new branch]]         
                                                                                                        action summary:                                                                                           
  publish (notneeded: 2, ok: 2)
```

Now, get a clone from github:

```
❱ datalad clone https://github.com/bpoldrack/public-testrepo clone-from-github
[INFO   ] scanning for unlocked files (this may take some time)                                          
[INFO   ] Remote origin not usable by git-annex; setting annex-ignore                                    
install(ok): /home/ben/playground/clone-from-github (dataset)
❱ cd clone-from-github 
❱ cat .git/config 
... <snip> ...
[remote "origin"]
	url = https://github.com/bpoldrack/public-testrepo
	fetch = +refs/heads/*:refs/remotes/origin/*
	annex-ignore = true
... <snip> ...
[remote "gin-sr"]
	url = https://gin.g-node.org/bpoldrack/public-test-repo2
	fetch = +refs/heads/*:refs/remotes/gin-sr/*
❱ datalad get some
get(ok): some (file) [from gin-sr...]                                                                    
```

**Comment by bpoldrack** [comment]
So, ATM I think two things may be needed:

1. `siblings` needs to check a git remote does in fact have an annex-uuid in order to allow for `--as-common-datasrc` since `git annex initremote` does not fail in this case, but creates an entry it would then not consider. Arguably, that could be considered a bug on annex' end. However, https://git-annex.branchable.com/special_remotes/git/#comment-bf19649e905fe1e5aed2ba283b2e1129 does not suggest Joey came to that conclusion.

2. I'd argue that the `create-sibling-*` commands should actually not just set up an unrelated repo on the respective service, add a git remote and be done, but do an `annex sync` or even `push` in order to make sure it's actually usable (before considering `--as-common-datasrc`). They do need a connection anyway (there's no offline scenario like with a simple `siblings add`) and I don't see why one would want to only set up a repo with unrelated history, no git-annex branch, no uuid and be done. Any scenario in which it would be useful to not have those basics set up as well, @datalad/developers ?

**Comment by mih** [comment]
> 1. `siblings` needs to check a git remote does in fact have an annex-uuid in order to allow for `--as-common-datasrc` since `git annex initremote` does not fail in this case, but creates an entry it would then not consider.

I can easily get behind that. We already do a URL protocol check. It can easily be extended.

> 2. I'd argue that the `create-sibling-*` commands should actually not just set up an unrelated repo on the respective service, add a git remote and be done, but do an `annex sync` or even `push` in order to make sure it's actually usable (before considering `--as-common-datasrc`). They do need a connection anyway (there's no offline scenario like with a simple `siblings add`) and I don't see why one would want to only set up a repo with unrelated history, no git-annex branch, no uuid and be done.

I don't see how this helps on top of (1). `create-sibling-ghlike` and friends have no `--as-common-datasrc`. They do not need to test for a scenario where someone would later *possibly* do that. Moreoever, wrapping `push` requires to expose all of the `push` API -- that is not sensible (I'd rather discontinue the idea of `--as-common-datasrc` entirely than do that. Also the online/offline argument does matter here, I think. When `siblings` fails to confirm a UUID locally, it can simply message users to run `update`. Not everything has to be fully automatic under all circumstances.

> I don't see why one would want to only set up a repo with unrelated history, no git-annex branch, no uuid and be done

That is not what `sync` or `push` do. They essentially make the remote match the local repo. There is a whole spectrum of use cases between that and what you described.

So in summary: I do not see why the entire fix could not be (1) -- to catch unsupported conditions, and the rest is adjusting documentation.

**Comment by bpoldrack** [comment]
>  create-sibling-ghlike and friends have no --as-common-datasrc. They do not need to test for a scenario where someone would later possibly do that.

That's true. But: We create a sibling for an annex on a platform that does support annex, but that git remote ends up being `annex-ignore`'ed, which means even a full `push` (with data) isn't easily possible right after `create-sibling-gin`. And in opposition to `as-common-datasrc`, a `push` is something one would most certainly want to do. What `create-sibling-gin` leaves behind is not ready to use.

> ... wrapping push ...
> That is not what sync or push do.

I didn't mean an `annex sync` or `datalad push` necessariy w/o any options. Just what we would possibly agree on needing. Could be something like `annex sync --only-annex --no-pull --no-content` or `push` with whatever specific options we consider needed - not a full wrapping.

Ultimately, I am kinda fine with only (1) - I just don't think the resulting workflow with that is very user-friendly, so I am exploring what we can reasonably do.
May be an alternative suggestion:
If `siblings add/configure` is called with `--as-common-datasrc`, *and* it figures that this git remote has no annex-uuid and/or is annex-ignored, it could simply try whether that's still appropriate and unset that `annex.ignore` + try to `enableremote`. If that still results in annex.ignore -> fail. If not proceed with `initremote` for the type-git special remote. WDYT about this, @mih ?

**Comment by mih** [comment]
> If siblings add/configure is called with --as-common-datasrc, and it figures that this git remote has no annex-uuid and/or is annex-ignored, it could simply try whether that's still appropriate and unset that annex.ignore + try to enableremote. If that still results in annex.ignore -> fail. If not proceed with initremote for the type-git special remote. WDYT about this, @mih ?

I think this all sounds nice on paper. But it comes with costs too. `siblings` will need communication with a service (generally not needed before). Given that the command's behavior is largely undefined now, this does not "break" things, but it increases complexity. I message asking to `update` avoids this. The same spirit of "just detect things automatically and make the right decisions" is leading to the problem in the first place (git-annex auto-probes exactly to determine what is "right", and does the "wrong" thing). IMHO user-friendly is not that it works somethings automatically when expectations magically align, but that it is comprehensibly what commands do (and not do), that the error messaging is clear and actionable. IMHO adjusting automagic behavior when it doesn't much is much harder and unfriendly than adding a simple extra step. My 2cts

**Comment by mih** [comment]
It would be good if this explorative issue could be replaced by concrete issues on the various aspects discovered here, broken down into actionable units. I started with #6209.

**Comment by mslw** [comment]
I can confirm that I was able to reproduce the steps described by @bpoldrack to set up a functioning common data source.

**Comment by bpoldrack** [comment]
Thx, @mslw!

FTR:
Regarding using a git-type special remote for a local location: I did not yet manage to make that work. Although I do have an annex-enabled, local git remote, the respective `git annex initremote` call does keep complaining `could not find existing git remote with specified location`.

**Comment by mih** [comment]
https://github.com/datalad/datalad/pull/6230 fixes the first part of this issue.

**Comment by mslw** [comment]
Thank you both!

> #6230 fixes the first part of this issue. Test now also includes an actual data-push to GIN

Not sure if the right place to answer is here or in the PR, so replying here due to context. I checked out the PR and quickly tried to repeat the `--as-common-datasrc` scenario outlined above, to see how it's affected. Looking at the command output and the `.git/config` file, with the proposed changes:
1. On creation (`datalad create-sibling-gin ...`) the remote has `annexurl`, no `annex-uuid`, and no `annex-ignore` config values.
2. This means the step `git config --unset-all remote.public-gin.annex-ignore` can be omitted :+1:.
3. The initial `datalad push ...` is still required though to obtain the `annex-uuid` (without it `siblings configure` fails, at least on a public repo).
4. After the push, `datalad siblings configure ... --as-common-datasrc ...` can be run.

> FTR:
Regarding using a git-type special remote for a local location: I did not yet manage to make that work. Although I do have an annex-enabled, local git remote, the respective git annex initremote call does keep complaining could not find existing git remote with specified location.

Thanks for looking into those - although now I also see that this is not the primarily intended use case for remotes, so perhaps it isn't much of a problem?

**Comment by adswa** [comment]
> This means the step git config --unset-all remote.public-gin.annex-ignore can be omitted +1.

wohoo! noted as a handbook todo, but will wait until this is resolved completely.

**Comment by bpoldrack** [comment]
> Thanks for looking into those - although now I also see that this is not the primarily intended use case for remotes, so perhaps it isn't much of a problem?

Yes, I think it's rare to want that to begin with. But still strange that it doesn't work as expected. Not clear to me, whether that's intended. Should be less of a security concern than `file:` URLs for files and I can't think of another reason to not allow for it. Will file an issue with git-annex.

**Comment by mslw** [comment]
> It would be good if this explorative issue could be replaced by concrete issues on the various aspects discovered here, broken down into actionable units. I started with #6209.

With what we've learned, I tried to revisit the "using GIN as a common data source" scenario and see where it currently stands. I think these problems are still open.

1. Push operation is needed before a GIN repo can be configured as a data source. Otherwise, the special remote gets configured without an annex UUID and things fail later on. From what I understand, we are currently leaning against including additional steps (requiring communication and potentially expensive) in the siblings command. If that is the case, then existing documentation needs to be updated - e.g. the docstring for `create_sibling_gin` suggests using `siblings configure` right after `create-sibling-gin` https://github.com/datalad/datalad/blob/6b4d5567eef9cef170b09895148a8cd991bea4ec/datalad/distributed/create_sibling_gin.py#L96-L101
2. Error behaviour when trying to configure a gin remote as common data source before pushing is different for public and private repositories. This already has a separate issue #6199 where I described it for `siblings add` (and didn't yet fully realise the problem - will add to that). To recap, for a public repo, `siblings configure` displays a git-annex error "could not find existing git remote with specified location"; for a private repo no such error is shown. Importantly, both cases end with a misconfigured remote. In light of the first point, both cases should probably end with a clear error or at least an explanatory message. I can't see why the private / public distinction causes different behaviour though.
3. The `siblings add ... --as-common-datasrc` is meaningful only for repos with an existing annex. In this thread we arrived at `create-sibling-gin` - `push` - `siblings configure ... --as-common-datasrc` as the right sequence of steps. However, the repo does not have to be created through a DataLad command. When a new gin repo is created manually, an analogous sequence of steps would be `siblings add` - unset annex-ignore - `push` - `siblings configure ... --as-common-datasrc`. For the reason described in the first point, the `--as-common-datasrc` must not be given to `siblings add`, although it is tempting (must not if the repo is clean; for an already populated repo this would be valid). Question is, whether it requires a change in `siblings` or just a documentation update (`create-siblings-gin` is yet unreleased, and the handbook currently suggests combining `siblings add` with `--as-common-datasrc`).

Should I open separate issues for 1 and 3?
